src/fluree/db/nameservice/core.cljc

```diff
@@ -0,0 +1,91 @@
+(ns fluree.db.nameservice.core
+  (:refer-clojure :exclude [exists?])
+  (:require [clojure.string :as str]
+            [fluree.db.conn.proto :as conn-proto]
+            [fluree.db.nameservice.proto :as ns-proto]
+            [fluree.db.util.async :refer [<? go-try]]
+            [clojure.core.async :as async :refer [go]]
+            [fluree.db.util.log :as log]))
+
+#?(:clj (set! *warn-on-reflection* true))
+
+(defn nameservices
+  [conn]
+  (conn-proto/-nameservices conn))
+
+(defn relative-ledger-alias?
+  [ledger-alias]
+  (not (str/starts-with? ledger-alias "fluree:")))
+
+(defn ns-address
+  "Returns async channel"
+  [nameservice ledger-alias branch]
+  (ns-proto/-address nameservice ledger-alias {:branch branch}))
+
+(defn addresses
+  "Retrieve address for each nameservices based on a relative ledger-alias.
+  If ledger-alias is not relative, returns only the current ledger alias.
+
+  TODO - if a single non-relative address is used, and the ledger exists,
+  we should retrieve all stored ns addresses in the commit if possible and
+  try to use all nameservices."
+  [conn ledger-alias {:keys [branch] :or {branch "main"} :as _opts}]
+  (if (relative-ledger-alias? ledger-alias)
+    (let [nameservices (nameservices conn)]
+      (go-try
+        (loop [nameservices* nameservices
+               addresses     []]
+          (let [ns (first nameservices*)]
+            (if ns
+              (if-let [address (<? (ns-address ns ledger-alias branch))]
+                (recur (rest nameservices*) (conj addresses address))
+                (recur (rest nameservices*) addresses))
+              addresses)))))
+    (go [ledger-alias])))
+
+(defn primary-address
+  "From a connection, lookup primary address from
+  nameservice(s) for a given ledger alias"
+  [conn ledger-alias opts]
+  (go-try
+    (first (<? (addresses conn ledger-alias opts)))))
+
+(defn push!
+  "Executes a push operation to all nameservices registered on the connection."
+  [conn commit-data]
+  (let [nameservices (nameservices conn)]
+    (go-try
+      (loop [nameservices* nameservices]
+        (when-let [ns (first nameservices*)]
+          (let [sync? (ns-proto/-sync? ns)]
+            (if sync?
+              (<? (ns-proto/-push ns commit-data))
+              (ns-proto/-push ns commit-data))
```

Comment:
So if I understand correctly, are you saying that instead of instantiating a `conn` with a list of `nameservers` as an option (one of which should be synchronous), that we have two different option parameters on the `conn`, one called something like `:synchronous-nameserver` and one `:asynchronous-nameservers`, and then enforce the sync behavior based on what bucket the nameserver was placed into? If so, that's a fairly simple change.

Reply:
Right. and then we wouldn't need the `sync?` method on the nameserver protocol and the protocol would be more generic. Information about whether the nameserver is synchronous is in relation to the conn, not independent of it.

Reply:
Another thought, if a generic nameservice (and its corresponding ns type/method) can just live in the commit as metadata, then it perhaps doesn't need to be "hard coded" into the conn. The conn can have a set of nameservices it knows, but the individual ledger/commit is configured to write to specific ones - one of which is "primary" and always asynchronous.

In this case, if you attempted to transact where the latest commit's nameservices are ones you know you can update, it will succeed - else it will throw an exception without you first changing the commit nameservices - not unlike how we allow defaultContext to get updated from time to time.

Reply:
Nameservices baked into the commit is still stretching my mental model and I'm still getting used to it, but I think I like this idea. It seems more flexible.